### PR TITLE
Feat: support SMAX

### PR DESCRIPTION
**Pull request by OmaymaMahjoub**
## What?
Support SMAX environment from [jaxmarl](https://github.com/FLAIROx/JaxMARL/tree/main)
## Why?
Add more environments that Mava supports
## How?
- Get action masking if exists
- Get global state instead of concatenate obs for centralized training.
- Get win rate during evaluation and log it to Neptune, json file, and cout in terminal
## Extra
TODO: add `27m_vs_30m` results to the readme

### CODE REVIEW

File: mava/wrappers/jaxmarl.py [hunk 1]
```diff
@@ -157,16 +157,23 @@ def __init__(self, env: MultiAgentEnv, timelimit: int = 500):
 
         self.agents = list(self._env.observation_spaces.keys())
 
+        # Action mask fn
+        if hasattr(self._env, "get_avail_actions"):
+            self.get_avail_actions = self._get_avail_actions
+        else:
+            self.get_avail_actions = lambda _: jnp.ones(self._action_shape)
```
Comment (sash-a) [inline mava/wrappers/jaxmarl.py]: `hasattr` is not something we want to do, do the jaxmarl envs not have a consistent way to get the action mask?

File: mava/wrappers/jaxmarl.py [hunk 2]
```diff
@@ -157,16 +157,23 @@ def __init__(self, env: MultiAgentEnv, timelimit: int = 500):
...
 
+        extras = {"global_state": jnp.array(obs["world_state"])}
         obs = Observation(
             agents_view=batchify(obs, self.agents),
-            action_mask=jnp.ones(self._action_shape),
+            action_mask=self.get_avail_actions(state),
             step_count=jnp.zeros(self._env.num_agents, dtype=int),
         )
-        return JaxMarlState(state, key, 0), restart(obs, extras={}, shape=(self.num_agents,))
+        return JaxMarlState(state, key, 0), restart(obs, extras=extras, shape=(self.num_agents,))
```
Comment (sash-a) [inline mava/wrappers/jaxmarl.py]: Lets not put global state in extras, we have and `ObservationGlobalState` so lets always use that for the global state and if there is no global state then we can just put a `jnp.zeros(1)` there. So then JaxMarl envs will always return this global state.

Only extra thing we will need to do is change the observation spec in this class a bit.

File: mava/wrappers/jaxmarl.py [hunk 3]
```diff
@@ -176,6 +183,7 @@ def step(
         obs, env_state, reward, done, infos = self._env.step(
             step_key, state.state, unbatchify(action, self.agents)
         )
+        infos["global_state"] = jnp.array(obs["world_state"])
```
Comment (sash-a) [inline mava/wrappers/jaxmarl.py]: Same here for global state

File: mava/wrappers/jaxmarl.py (hunk 1 above)
Comment (sash-a) [inline mava/wrappers/jaxmarl.py]: If we can't get rid of the `hasattr` this is my suggestion
```suggestion
        # True if the wrapped env has a function to get the action mask
        has_action_mask = hasattr(self._env, "get_avail_actions")
```

File: mava/wrappers/jaxmarl.py [hunk 4]
```diff
@@ -226,3 +234,8 @@ def discount_spec(self) -> specs.BoundedArray:
         return specs.BoundedArray(
             shape=(self.num_agents,), dtype=float, minimum=0.0, maximum=1.0, name="discount"
         )
+
+    def _get_avail_actions(self, state: JaxMarlState) -> Array:
+        avail_actions = self._env.get_avail_actions(state)
+        action_mask = jnp.array(batchify(avail_actions, self.agents), dtype=jnp.float32)
+        return action_mask
```
Comment (sash-a) [inline mava/wrappers/jaxmarl.py]:
```suggestion
    def action_mask(self, state: JaxMarlState) -> Array:
        if self.has_action_mask:
          avail_actions = self._env.get_avail_actions(state)
          mask = jnp.array(batchify(avail_actions, self.agents), dtype=jnp.float32)
        else:
            mask = jnp.ones(self._action_shape)

        return mask
```

File: mava/wrappers/jaxmarl.py [hunk 5]
```diff
@@ -157,16 +157,23 @@ def __init__(self, env: MultiAgentEnv, timelimit: int = 500):
...
+
     def reset(self, key: PRNGKey) -> Tuple[JaxMarlState, TimeStep[Observation]]:
         key, reset_key = jax.random.split(key)
         obs, state = self._env.reset(reset_key)
 
+        extras = {"global_state": jnp.array(obs["world_state"])}
         obs = Observation(
             agents_view=batchify(obs, self.agents),
-            action_mask=jnp.ones(self._action_shape),
+            action_mask=self.get_avail_actions(state),
```
Comment (sash-a) [inline mava/wrappers/jaxmarl.py]:
```suggestion
            action_mask=self.action_mask(state),
```

File: mava/wrappers/jaxmarl.py [hunk 6]
```diff
@@ -157,16 +157,23 @@ def __init__(self, env: MultiAgentEnv, timelimit: int = 500):
...
+        if hasattr(self._env, "get_avail_actions"):
+            self.get_avail_actions = self._get_avail_actions
+        else:
+            self.get_avail_actions = lambda _: jnp.ones(self._action_shape)
+
     def reset(self, key: PRNGKey) -> Tuple[JaxMarlState, TimeStep[Observation]]:
         key, reset_key = jax.random.split(key)
         obs, state = self._env.reset(reset_key)
 
+        extras = {"global_state": jnp.array(obs["world_state"])}
```
Comment (sash-a) [inline mava/wrappers/jaxmarl.py]: Use `ObservationGlobalState`
```suggestion
```

File: mava/wrappers/shared.py [hunk 7]
```diff
@@ -142,40 +142,39 @@ class GlobalStateWrapper(Wrapper):
...
-        global_state = jnp.concatenate(timestep.observation.agents_view, axis=0)
-        global_state = jnp.tile(global_state, (self._env.num_agents, 1))
-
-        observation = ObservationGlobalState(
-            global_state=global_state,
-            agents_view=timestep.observation.agents_view,
-            action_mask=timestep.observation.action_mask,
-            step_count=timestep.observation.step_count,
-        )
+    def __init__(self, env: Environment, add_agent_id: bool = False) -> None:
```
Comment (sash-a) [inline mava/wrappers/shared.py]: I think we should leave this class untouched and just use the `ObservationGlobalState` directly in the `JaxMarl` wrapper, let me know if you think that will cause issues?

File: mava/utils/logger_tools.py [hunk 8]
```diff
@@ -140,10 +145,14 @@ def get_neptune_logger(cfg: Dict) -> neptune.Run:
 
 def get_experiment_path(config: Dict, logger_type: str) -> str:
     """Helper function to create the experiment path."""
+    task_name = (
+        config["env"]["scenario"]
+        if isinstance(config["env"]["scenario"], str)
+        else config["env"]["scenario"]["task_name"]
```
Comment (sash-a) [inline mava/utils/logger_tools.py]: Is it possible to unify this in the config so that we don't need this `if`?

File: mava/utils/logger_tools.py [hunk 9]
```diff
@@ -77,11 +78,15 @@ def _setup_json(self, cfg: Dict) -> None:
             json_logs_path = os.path.join(
                 cfg["logger"]["base_exp_path"], "json", cfg["logger"]["kwargs"]["json_path"]
             )
-
+        task_name = (
+            cfg["env"]["scenario"]
+            if isinstance(cfg["env"]["scenario"], str)
+            else cfg["env"]["scenario"]["task_name"]
```
Comment (sash-a) [inline mava/utils/logger_tools.py]: Same here

File: mava/evaluator.py [hunk 10]
```diff
@@ -262,7 +277,7 @@ def evaluator_setup(
     """Initialise evaluator_fn."""
     # Get available TPU cores.
     n_devices = len(jax.devices())
-
+    win_rate = config["env"]["env_name"] == "HeuristicEnemySMAX"
```
Comment (sash-a) [inline mava/evaluator.py]: Slightly widens this condition if we use the other envs in the future

```suggestion
    win_rate = isinstance(env, (SMAX, HeuristicEnemySMAX, LearnedPolicyEnemySMAX))
```

File: mava/logger.py [hunk 11]
```diff
@@ -58,6 +58,11 @@ def log(
         if absolute_metric:
             prefix = "absolute/"
             episodes_info = metrics.episodes_info
+            # Add win rate to episodes_info in case it exists.
+            if "won_episode" in episodes_info:
+                win_rate = (
+                    jnp.sum(episodes_info["won_episode"]) / (logger.num_eval_episodes * 10)
+                ) * 100
```
Comment (sash-a) [inline mava/logger.py]: nit
```suggestion
                n_won_episodes = jnp.sum(episodes_info["won_episode"])
                n_episodes = logger.num_eval_episodes * 10
                win_rate = (n_won_episodes / n_episodes) * 100
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Awesome work @OmaymaMahjoub! 

Main thing I think we should do is use the `ObservationGlobalState` directly, that should reduce the amount of changes a lot!

File: mava/wrappers/jaxmarl.py (hunk 1 above)
Comment (OmaymaMahjoub) [reply mava/wrappers/jaxmarl.py]: unfortunately, based on what I found I don't think there is a standard way to have actions mask, for exampla smax uses `get_avail_actions`, hanabi uses `get_legal_moves`  and mabrax doesn't have the method as it is continuous actions

Comment (OmaymaMahjoub) [reply mava/wrappers/jaxmarl.py]: shall we keep it `_get_avail_actions` and solve this issue another time when we support more envs from jaxmarl?

Comment (sash-a) [reply mava/wrappers/jaxmarl.py]: Ye this really sucks, keep it for now then, I think we'll have to open an issue on their repo

File: mava/wrappers/shared.py [hunk 12]
```diff
@@ -120,11 +120,14 @@ def observation_spec(
         agents_view = specs.Array(
             (self._env.num_agents, num_obs_features), jnp.int32, "agents_view"
         )
+
+        if hasattr(obs_spec, "global_state"):
+            state_size = obs_spec.global_state.shape[-1] + self._env.num_agents
+        else:
+            state_size = num_obs_features * self._env.num_agents + self._env.num_agents
```
Comment (sash-a) [inline mava/wrappers/shared.py]: I think we can collapse this condition into `self.has_global_state` we must just make sure has_global_state is always true for `JaxMarlEnvs`. And then also use your way of calculating state size (with the `obs_spec`) as it's more future proof.


```python
if self.has_global_state:
    wrapped_state_shape = obs_spec.global_state.shape
    state_shape = (*wrapped_state_shape[:-1], wrapped_state_shape[-1] + self._env.num_agents)
    global_state = specs.Array(state_shape, ...)
    return obs_spec.replace(agents_view=agents_view, global_state=global_state)

return obs_spec.replace(agents_view=agents_view)
```

File: mava/wrappers/jaxmarl.py [hunk 13]
```diff
@@ -163,36 +163,36 @@ def __init__(self, env: MultiAgentEnv, timelimit: int = 500):
...
-        extras = {"global_state": jnp.array(obs["world_state"])}
-        obs = Observation(
+        obs = ObservationGlobalState(
             agents_view=batchify(obs, self.agents),
             action_mask=self.get_avail_actions(state),
+            global_state=self.get_global_state(obs),
             step_count=jnp.zeros(self._env.num_agents, dtype=int),
         )
-        return JaxMarlState(state, key, 0), restart(obs, extras=extras, shape=(self.num_agents,))
+        return JaxMarlState(state, key, 0), restart(obs, extras={}, shape=(self.num_agents,))
```
Comment (sash-a) [inline mava/wrappers/jaxmarl.py]: extras default is `{}`
```suggestion
        return JaxMarlState(state, key, 0), restart(obs, shape=(self.num_agents,))
```

File: mava/utils/make_env.py [hunk 14]
```diff
@@ -100,8 +100,9 @@ def make_jaxmarl_env(env_name: str, config: Dict) -> Tuple[Environment, Environm
     env = JaxMarlWrapper(jaxmarl.make(env_name, **kwargs))
     eval_env = JaxMarlWrapper(jaxmarl.make(env_name, **kwargs))
 
-    env = add_optional_wrappers(env, config)
-    eval_env = add_optional_wrappers(eval_env, config)
+    if config["system"]["add_agent_id"]:
+        env = AgentIDWrapper(env, config["system"]["add_global_state"])
+        eval_env = AgentIDWrapper(eval_env, config["system"]["add_global_state"])
```
Comment (sash-a) [inline mava/utils/make_env.py:108]: If you like the changes suggested below then just make these both `True` and put a comment that JaxMarlEnvs always have global state

**Review by sash-a — CHANGES_REQUESTED** [review]
Nice changes! Just a couple more minor things :pray:

File: mava/evaluator.py (hunk 10 above)
Comment (OmaymaMahjoub) [reply mava/evaluator.py]: The issue here is that env is not an instance of SMAX... as it is a wrapped env therefore this `isinstance` will be very nested by mentioning `env._env._env...` depending on how many wrappers are added

File: mava/wrappers/jaxmarl.py (hunk 13 above)
Comment (OmaymaMahjoub) [reply mava/wrappers/jaxmarl.py]: The default is None which causes an error for the type change:
```
TypeError: Scanned function carry input and carry output must have the same pytree structure, but they differ:
  * the input carry component learner_state.timestep.extras is a <class 'NoneType'> but the corresponding component of the carry output is a <class 'dict'>, so their Python types differ

Revise the scanned function so that its output is a pair where the first element has the same pytree structure as the first argument.
```

File: mava/evaluator.py [hunk 15]
```diff
@@ -86,6 +90,11 @@ def not_done(carry: Tuple) -> bool:
             "episode_return": final_state.return_,
             "episode_length": final_state.step_count_,
         }
+        # Log won episode if win rate is required.
+        if win_rate:
+            eval_metrics["won_episode"] = jnp.where(
+                jnp.all(final_state.timestep.reward >= 1.0), 1, 0
+            )
```
Comment (sash-a) [inline mava/evaluator.py]: Slightly cleaner

```suggestion
            eval_metrics["won_episode"] = jnp.all(final_state.timestep.reward >= 1.0).astype(int)
```

File: mava/evaluator.py [hunk 16]
```diff
@@ -192,6 +202,11 @@ def not_done(carry: Tuple) -> bool:
...
```
Comment (sash-a) [inline mava/evaluator.py]: same here

```suggestion
            eval_metrics["won_episode"] = jnp.all(final_state.timestep.reward >= 1.0).astype(int)
```

File: mava/utils/logger_tools.py [hunk 17]
```diff
@@ -51,6 +51,7 @@ def __init__(self, cfg: Dict) -> None:
         self.should_log = bool(
             cfg["logger"]["use_json"] or cfg["logger"]["use_tf"] or cfg["logger"]["use_neptune"]
         )
+        self.num_eval_episodes = cfg["arch"]["num_eval_episodes"]
```
Comment (sash-a) [inline mava/utils/logger_tools.py:54]: Why is this here? Doesn't seem like we use it

**Review by sash-a — DISMISSED** [review]
Minor style change otherwise I'm super happy with this, great job!

File: mava/logger.py [hunk 18]
```diff
@@ -58,6 +58,11 @@ def log(
         if absolute_metric:
             prefix = "absolute/"
             episodes_info = metrics.episodes_info
+            # Add win rate to episodes_info in case it exists.
+            if "won_episode" in episodes_info:
+                n_won_episodes = jnp.sum(episodes_info["won_episode"])
+                n_episodes = logger.num_eval_episodes * 10
```
Comment (callumtilbury) [inline mava/logger.py:68]: Could we please explain why *10 here?

File: mava/evaluator.py (hunk 15 above)
Comment (callumtilbury) [reply mava/evaluator.py]: +1

File: mava/evaluator.py [hunk 19]
```diff
@@ -262,7 +277,7 @@ def evaluator_setup(
     """Initialise evaluator_fn."""
     # Get available TPU cores.
     n_devices = len(jax.devices())
-
+    win_rate = config["env"]["env_name"] in ["HeuristicEnemySMAX", "LearnedPolicyEnemySMAX"]
```
Comment (callumtilbury) [inline mava/evaluator.py]: Small explanation here might be helpful, especially because the variable may be seen as "the win rate" itself.

File: mava/evaluator.py [hunk 20]
```diff
@@ -275,28 +290,27 @@ def evaluator_setup(
...
         vmapped_eval_network_apply_fn = jax.vmap(
             network.apply,
             in_axes=(None, 0),
         )
-        evaluator = get_ff_evaluator_fn(
-            eval_env,
-            vmapped_eval_network_apply_fn,
-            config,
-        )
+        evaluator = get_ff_evaluator_fn(eval_env, vmapped_eval_network_apply_fn, config, win_rate)
```
Comment (callumtilbury) [inline mava/evaluator.py]: Perhaps same formatting as the others
```suggestion
        evaluator = get_ff_evaluator_fn(
            eval_env,
            vmapped_eval_network_apply_fn,
            config,
            win_rate,
        )
```

File: mava/utils/logger_tools.py (hunk 17 above)
Comment (callumtilbury) [reply mava/utils/logger_tools.py:54]: It seems to be used here as a public member variable:
https://github.com/instadeepai/Mava/blob/771283f2327718248203d4b64728c3f77453e0f0/mava/logger.py#L64

File: mava/wrappers/jaxmarl.py [hunk 21]
```diff
@@ -142,7 +142,7 @@ def jaxmarl_space_to_jumanji_spec(space: jaxmarl_spaces.Space) -> specs.Spec:
 class JaxMarlWrapper(Wrapper):
     """Wraps a JaxMarl environment so that its API is compatible with jumaji environments."""
```
Comment (callumtilbury) [inline mava/wrappers/jaxmarl.py]:
```suggestion
    """Wraps a JaxMarl environment so that its API is compatible with Jumanji environments."""
```

File: mava/utils/logger_tools.py (hunk 17 above)
Comment (OmaymaMahjoub) [reply mava/utils/logger_tools.py:54]: Because we need it to calculate the win rate in logger.py: `win_rate = (n_won_episodes / logger.num_eval_episodes) * 100`

File: mava/utils/make_env.py [hunk 22]
```diff
@@ -94,14 +94,17 @@ def make_jaxmarl_env(env_name: str, config: Dict) -> Tuple[Environment, Environm
 
     kwargs = config["env"]["kwargs"]
     if "smax" in env_name.lower():
-        kwargs["scenario"] = map_name_to_scenario(config["env"]["scenario"])
+        kwargs["scenario"] = map_name_to_scenario(config["env"]["scenario"]["task_name"])
 
     # Placeholder for creating JAXMARL environment.
```
Comment (callumtilbury) [inline mava/utils/make_env.py]: Is this still just a 'placeholder'?

**Review by sash-a — DISMISSED** [review]
:rocket:

File: mava/wrappers/jaxmarl.py [hunk 23]
```diff
@@ -226,3 +262,16 @@ def discount_spec(self) -> specs.BoundedArray:
...
             shape=(self.num_agents,), dtype=float, minimum=0.0, maximum=1.0, name="discount"
         )
+
+    def action_mask(self, state: JaxMarlState) -> Array:
+        """Get action mask for each agent."""
+        if self.has_action_mask:
+            avail_actions = self._env.get_avail_actions(state)
+            mask = jnp.array(batchify(avail_actions, self.agents), dtype=jnp.float32)
+        else:
+            mask = jnp.ones(self._action_shape)
```
Comment (callumtilbury) [inline mava/wrappers/jaxmarl.py]:
```suggestion
            mask = jnp.ones(self._action_shape, dtype=jnp.float32)
```
Just to be explicit + consistent with the line above :)

**Review by callumtilbury — COMMENTED** [review]
A few minor points, but otherwise looks good :) Thanks @OmaymaMahjoub!

File: mava/logger.py (hunk 18 above)
Comment (callumtilbury) [reply mava/logger.py:68]: (As discussed privately 👍🏻 )

File: mava/evaluator.py (hunk 19 above)
Comment (callumtilbury) [reply mava/evaluator.py]: Or rename to e.g. `include_win_rate`?

File: mava/utils/make_env.py (hunk 22 above)
Comment (callumtilbury) [reply mava/utils/make_env.py]: Sorry, this is pointing to the wrong location. See line 99 below 👍🏻

Comment (OmaymaMahjoub) [reply mava/utils/make_env.py]: I updated the comment to be more informative

File: mava/evaluator.py [hunk 24]
```diff
@@ -35,7 +35,7 @@ def get_ff_evaluator_fn(
     env: Environment,
     apply_fn: ActorApply,
     config: dict,
-    win_rate: bool = False,
+    include_win_rate: bool = False,
```
Comment (sash-a) [inline mava/evaluator.py]: I like being more explicit here, sorry one final thing: 
```suggestion
    log_win_rate: bool = False,
```

File: mava/evaluator.py [hunk 25]
```diff
@@ -91,7 +91,7 @@ def not_done(carry: Tuple) -> bool:
             "episode_length": final_state.step_count_,
         }
         # Log won episode if win rate is required.
-        if win_rate:
+        if include_win_rate:
```
Comment (sash-a) [inline mava/evaluator.py]:
```suggestion
        if log_win_rate:
```

File: mava/evaluator.py [hunk 26]
```diff
@@ -131,7 +131,7 @@ def get_rnn_evaluator_fn(
     apply_fn: RecActorApply,
     config: dict,
     scanned_rnn: nn.Module,
-    win_rate: bool = False,
+    include_win_rate: bool = False,
```
Comment (sash-a) [inline mava/evaluator.py]:
```suggestion
    log_win_rate: bool = False,
```

File: mava/evaluator.py [hunk 27]
```diff
@@ -201,7 +201,7 @@ def not_done(carry: Tuple) -> bool:
             "episode_length": final_state.step_count_,
         }
         # Log won episode if win rate is required.
-        if win_rate:
+        if include_win_rate:
```
Comment (sash-a) [inline mava/evaluator.py]:
```suggestion
        if log_win_rate:
```

File: mava/evaluator.py [hunk 28]
```diff
@@ -273,40 +273,40 @@ def evaluator_setup(
     """Initialise evaluator_fn."""
     # Get available TPU cores.
     n_devices = len(jax.devices())
-    win_rate = config["env"]["env_name"] in ["HeuristicEnemySMAX", "LearnedPolicyEnemySMAX"]
+    # Check if win rate is required for evaluation.
+    include_win_rate = config["env"]["env_name"] in ["HeuristicEnemySMAX", "LearnedPolicyEnemySMAX"]
```
Comment (sash-a) [inline mava/evaluator.py]:
```suggestion
    log_win_rate = config["env"]["env_name"] in ["HeuristicEnemySMAX", "LearnedPolicyEnemySMAX"]
```

File: mava/evaluator.py [hunk 29]
```diff
@@ -273,40 +273,40 @@ def evaluator_setup(
...
         )
-        evaluator = get_ff_evaluator_fn(eval_env, vmapped_eval_network_apply_fn, config, win_rate)
+        evaluator = get_ff_evaluator_fn(eval_env, vmapped_eval_apply_fn, config, include_win_rate)
         absolute_metric_evaluator = get_ff_evaluator_fn(
             eval_env,
-            vmapped_eval_network_apply_fn,
+            vmapped_eval_apply_fn,
             config,
-            win_rate,
+            include_win_rate,
```
Comment (sash-a) [inline mava/evaluator.py]:
```suggestion
            log_win_rate,
```

File: mava/evaluator.py [hunk 30]
```diff
@@ -273,40 +273,40 @@ def evaluator_setup(
...
         )
         absolute_metric_evaluator = get_rnn_evaluator_fn(
             eval_env,
-            vmapped_eval_network_apply_fn,
+            vmapped_eval_apply_fn,
             config,
             scanned_rnn,
-            win_rate,
+            include_win_rate,
```
Comment (sash-a) [inline mava/evaluator.py]:
```suggestion
            log_win_rate,
```

File: mava/evaluator.py [hunk 31]
```diff
@@ -273,40 +273,40 @@ def evaluator_setup(
...
             network.apply, in_axes=(None, 1, (2, None)), out_axes=(1, 2)
         )
         evaluator = get_rnn_evaluator_fn(
             eval_env,
-            vmapped_eval_network_apply_fn,
+            vmapped_eval_apply_fn,
             config,
             scanned_rnn,
-            win_rate,
+            include_win_rate,
```
Comment (sash-a) [inline mava/evaluator.py]:
```suggestion
            log_win_rate,
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Really small issue I have with the new variable name, sorry! After this I'm happy to :ship: it

**Review by sash-a — APPROVED** [review]
:rocket:

**Review by callumtilbury — APPROVED** [review]
🚗
